demo/app.py

```diff
@@ -1,32 +1,48 @@
 import json
 import re
 import requests
-from flask import Flask, request, render_template
+import sys
+import xml.etree.ElementTree as elementTree
+from flask import Flask, request, render_template, redirect, session
 from requests.exceptions import ConnectionError
 
 
 app = Flask(__name__)
 app.config.from_pyfile("configuration.py")
 
 
-def init_session():
-    global session
-    session = requests.Session()
-    session.auth = (
-        app.config["AUTH_USERNAME"],
-        app.config["AUTH_PASSWORD"]
-    )
-    session.verify = False
-
-
 @app.route("/", methods=["GET", "POST"])
 def base():
+    cas_url = app.config["CAS_URL"]
+    self_url = request.url_root
+    authorized_users = app.config["AUTHORIZED_USERS"]
+
     if request.method == "POST":
         return get_response(request.form["search"])
+    elif "ticket" in request.args:
+        try:
+            user = validate_cas(cas_url, request.args["ticket"], self_url)
+        except CASError as e:
+            return "login failed {}".format(str(e)), 403
+        except Exception as e:
+            return "login failed", 403
+        session["user"] = user
+        return redirect(request.path)
+    elif "user" not in session:
+        return redirect(cas_url + "/login?service=" + self_url)
+    if session["user"] not in authorized_users:
+        return render_template("index.html", unauthorized=True)
```

Comment:
I realized this is still returning a `200` response code. It should probably be a `403`.